src/Listener/DoctrineMappingListener.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\UserBundle\Listener;
+
+use Doctrine\ORM\Event\LoadClassMetadataEventArgs;
+
+/**
+ * @internal
+ */
+final class DoctrineMappingListener
+{
+    private bool $isArrayTypeAvailable;
+
+    public function __construct(private string $userClass)
+    {
+        $this->isArrayTypeAvailable = class_exists('Doctrine\DBAL\Types\ArrayType');
+    }
+
+    /**
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function loadClassMetadata(LoadClassMetadataEventArgs $event): void
+    {
+        $metadata = $event->getClassMetadata();
+
+        if (!$this->isArrayTypeAvailable) {
+            return;
+        }
+
+        if ($metadata->getName() !== $this->userClass) {
+            return;
+        }
+
+        if (!$metadata->hasField('roles')) {
+            return;
+        }
+
+        /**
+         * @psalm-suppress InvalidPropertyAssignmentValue
+         */
+        $metadata->fieldMappings['roles']['type'] = 'array';
```

Comment:
Is it the right way to change the value ?

The property is documented as READ-ONLY
https://github.com/doctrine/orm/blob/3d9af3187f59930972e7fcb90a1d8059a37b8032/src/Mapping/ClassMetadata.php#L334-L339

Reply:
I think 
```
public function setAttributeOverride(string $fieldName, array $overrideMapping): void
```
should be used instead. (https://github.com/doctrine/orm/blob/3.1.x/src/Mapping/ClassMetadata.php#L1737)

Reply:
@VincentLanglet https://github.com/doctrine/orm/blob/3.1.x/src/Mapping/ClassMetadata.php#L1765-L1767 it is not possible to change the type using this way. 

This solution is a hack but to be honest having a second class `BaseUser3` just to change the type is not good either. This can end up, with `BaseUser4` etc ...

Reply:
Any recommendation for such situation @greg0ire ? Is it the best way ?